### PR TITLE
Add a cleanup handling interface to route agent

### PR DESCRIPTION
**Pull request by mangelajo**
This removes technical debt where specific knowledge
related to cable drivers was implemented in the route-agent.

Introduces an interface which can be exposed from multiple
places, for example can be used by [1] to improve the
separation of concerns in the code.

[1] https://github.com/submariner-io/submariner/pull/762

Co-authored-by: Sridhar Gaddam <sgaddam@redhat.com>

### CODE REVIEW

File: pkg/cable/driver.go [hunk 1]
```diff
@@ -4,10 +4,10 @@ import (
 	"fmt"
 	"strings"
 
-	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"k8s.io/klog"
```
Comment (mangelajo) [inline pkg/cable/driver.go:7]: ehhm, ok, this doesn't belong in this patch

File: pkg/routeagent/controllers/route/route.go [hunk 2]
```diff
@@ -681,11 +688,10 @@ func (r *Controller) processNextEndpoint() bool {
 		}
 		// If the active Gateway transitions to a new node, we flush the HostNetwork routing table.
 		r.updateRoutingRulesForHostNetworkSupport(nil, FlushRouteTable)
-		r.gwVxLanMutex.Unlock()
 
-		// NOTE(mangelajo): This may not belong here, it's a gateway cleanup thing
-		r.cleanStrongswanRoutingTable()
-		r.cleanXfrmPolicies()
+		r.GatewayToNonGatewayTransition()
```
Comment (mangelajo) [inline pkg/routeagent/controllers/route/route.go:692]: The order of the mutex has been modified after talking with @sridhargaddam as it seemed the safest

File: pkg/routeagent/cleanuphandlers/cleanuphandlers.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package cleanuphandlers
+
+type CleanupHandler interface {
```
Comment (tpantelis) [inline pkg/routeagent/cleanuphandlers/cleanuphandlers.go:3]:
```suggestion
type Handler interface {
```

This avoids stutter - usages will now read `cleanup.Handler.`

File: pkg/routeagent/cleanuphandlers/cleanuphandlers.go [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+	GetName() string
```
Comment (tpantelis) [inline pkg/routeagent/cleanuphandlers/cleanuphandlers.go:4]: I don't see this used anywhere ...

File: pkg/routeagent/controllers/route/cleanuphandling.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+func (r *Controller) InstallCleanupHandlers(handlers []cleanuphandlers.CleanupHandler) {
+	r.cleanupHandlers = append(r.cleanupHandlers, handlers...)
+}
+
+func (r *Controller) GatewayToNonGatewayTransition() {
+	for _, handler := range r.cleanupHandlers {
+		if err := handler.GatewayToNonGatewayTransition(); err != nil {
+			klog.Errorf("Error handling GatewayToNonGateway cleanup in %v, %s", handler, err)
```
Comment (tpantelis) [inline pkg/routeagent/controllers/route/cleanuphandling.go:15]: Is this where `GetName` was intended to be used?

```suggestion
			klog.Errorf("Error handling GatewayToNonGateway cleanup in %q, %s", handler.GetName(), err)
```

File: pkg/routeagent/controllers/route/cleanuphandling.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package route
+
+import (
+	"github.com/submariner-io/submariner/pkg/routeagent/cleanuphandlers"
+	"k8s.io/klog"
+)
+
+func (r *Controller) InstallCleanupHandlers(handlers []cleanuphandlers.CleanupHandler) {
```
Comment (tpantelis) [inline pkg/routeagent/controllers/route/cleanuphandling.go:8]:
```suggestion
func (r *Controller) installCleanupHandlers(handlers []cleanuphandlers.CleanupHandler) {
```

File: pkg/routeagent/controllers/route/cleanuphandling.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+	r.cleanupHandlers = append(r.cleanupHandlers, handlers...)
+}
+
+func (r *Controller) GatewayToNonGatewayTransition() {
```
Comment (tpantelis) [inline pkg/routeagent/controllers/route/cleanuphandling.go:12]:
```suggestion
func (r *Controller) gatewayToNonGatewayTransition() {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]
I also suggest to use '_' in file names to improve readability, eg _routecleanuphandler_ -> _route_cleanup_handler_.

File: pkg/cable/cleanups/defaultcleanups.go [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package cleanups
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply pkg/cable/cleanups/defaultcleanups.go:1]: But this has to be cleanups as for the other files in this directory, right? :-?

File: pkg/cable/cleanups/routecleanuphandler.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"os/exec"
+	"strconv"
+
+	"github.com/submariner-io/submariner/pkg/routeagent/cleanuphandlers"
+)
+
+type RoutingTableCleanupHandler struct {
+	cleanuphandlers.CleanupHandler
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply pkg/cable/cleanups/routecleanuphandler.go:11]: I thought it didn't even add a field, does it? I know it's optional. It was more about being explicit to the human reader to know we are implementing that interface, but we can remove it, it's obvious.

File: pkg/routeagent/cleanuphandlers/cleanuphandlers.go [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package cleanuphandlers
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply pkg/routeagent/cleanuphandlers/cleanuphandlers.go:1]: Ack, I will rename the directory too.

File: pkg/routeagent/cleanuphandlers/cleanuphandlers.go (hunk 4 above)
Comment (mangelajo) [reply pkg/routeagent/cleanuphandlers/cleanuphandlers.go:4]: Right, I wanted to use it on the error handling and forgot. Thanks

File: pkg/cable/cleanups/defaultcleanups.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/cable/cleanups/defaultcleanups.go]: > But this has to be cleanups as for the other files in this directory, right? :-?

 mean to change the package and directory to cleanup.

File: pkg/cable/cleanups/routecleanuphandler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/cable/cleanups/routecleanuphandler.go]: That does add a field. named `CleanupHandler`.  Try calling `handler. CleanupHandler.GetName()` - you'll get a panic due to nil  😄 .  Unfortunately there's no  built-in way to show what interfaces are implemented, other than a comment.

File: pkg/cable/cleanups/defaultcleanups.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mangelajo) [reply pkg/cable/cleanups/defaultcleanups.go]: Done. Sorry I didn't understand.

**Review by tpantelis — COMMENTED** [review]
Looks good. A nit would be to rename _default_cleanups.go_ ->  `default_cleanup.go` for consistency.

**Review by sridhargaddam — APPROVED** [review]
LGTM, thanks.

**Comment by mangelajo** [comment]
> Looks good. A nit would be to rename _default_cleanups.go_ -> `default_cleanup.go` for consistency.

That's too nitpicky ;-) it's actually multiple cleanups inside ;) but ok :-)

**Review by tpantelis — APPROVED** [review]

**Review by sridhargaddam — APPROVED** [review]
